### PR TITLE
feat: rule no-unsafe-values

### LINKED ISSUE #29: New Rule: unsafe values
### Rule details

Fail on values that aren't safe for interchange

### What type of rule is this?

Warns about a potential problem

### Example code

Per https://socket.dev/blog/judicious-json

Some values of JSON do not coerce safely across different situations.

1. Non-finite numbers. The can coerce to infinity or other precision loss.

```js
// 'null'
JSON.stringify(Number.MAX_VALUE * 10)
// Infinity
JSON.parse('2e309')
```

2. Lone surrogates. These actually throw errors in various libs.

```js
let serialized = JSON.stringify("🔥"[0]);
// '"\\uD83D"' - note extra \
JSON.parse(serialized);
// "\uD83D"
```

3. Array/string bounds. These have hard limits in some languages like JS.

This isn't feasible to show an example of. Only affects very large data. Uncommon and easily skipped usually.

### Participation

- [x] I am willing to submit a pull request to implement this rule.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by bmeck**
#### Prerequisites checklist

-   [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Introduce a new rule for checking common interchange problems

#### What changes did you make? (Give an overview)

Added a rule for:

- non-finite (Infinity, null, NaN) values.
- lone surrogates due to UTF16

#### Related Issues

fixes #29

#### Is there anything you'd like reviewers to focus on?

tests are somewhat limited by using already parsed values so I avoided trying to reparse for getting exact locations within strings and/or couldn't check array/string bounds since things error before getting to the rules processing phase.

### CODE REVIEW

**Comment by bmeck** [comment]
Easy CLA is giving me a wildly inaccurate corporate account. Will have to work around sometime as I can't select a different corporation

**Review by nzakas — CHANGES_REQUESTED** [review]
This looks like a good start. Some high-level notes:

1. Please add this rule to the rules table on the README
2. Please add this rule to the recommended configuration

Also, it looks like your last invalid test is actually producing two errors instead of one. Can you double-check that?

File: src/rules/no-unsafe-values.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+				// match any low surrogate not already matched
+				const surrogatePattern =
+					/[\uD800-\uDBFF][\uDC00-\uDFFF]?|[\uDC00-\uDFFF]/gu;
+				let match = surrogatePattern.exec(node.value);
+				while (match) {
+					// only need to report non-paired surrogates
+					if (match[0].length < 2) {
+						context.report({
+							loc: node.loc,
+							messageId: "loneSurrogate",
```
[the remark this answers is not in the document]
Comment (bmeck) [reply src/rules/no-unsafe-values.js:46]: sure

File: tests/rules/no-unsafe-values.test.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		{
+			code: '"\ud83d"',
+			errors: [
+				{
+					messageId: "loneSurrogate",
+					line: 1,
+					column: 1,
+					endLine: 1,
+					endColumn: 4,
+				},
```
Comment (mdjermanovic) [inline tests/rules/no-unsafe-values.test.js:73]: Can you add `data` to all test errors with `messageId: "loneSurrogate"`?

**Comment by nzakas** [comment]
@bmeck just one outstanding comment to address here.

**Review by nzakas — APPROVED** [review]
Because we're just waiting for a change to the tests, I'll merge this and then update the tests myself.

**Pull request merged**